Gestion des droits entre super-admin + GESARTICLE

mattgu74 opened the pull request:
Il est maintenant possible de set/remove/consult les droits super-admin (ou `fun_id = NULL`)

Une pull request arrive dans scoobydoo pour avoir l'interface qui va avec.

mattgu74 committed:
getFundations reconnait lorsque l'on est admin

Pour éviter la multiplication des requetes entre les clients et le serveur.
Il me semble judicieux de venir ajouter une clef pour "toutes les fundations"
Lorsque l'on est admin et que l'on récupére la liste des fundations.

mattgu74 committed:
NULL ne passe pas vraiment en json enfait, du coup ce sera "NULL"

mattgu74 committed:
Gestion des droits entre super-admin

Il est maintenant possible de set/remove/consult les droits super-admin (ou `fun_id = NULL`)

Une pull request arrive dans scoobydoo pour avoir l'interface qui va avec.

apuyou commented on class/ApplicationRight.class.php:
```diff
@@ -27,7 +27,7 @@ public static function check($application_id = false, $service_name = false, $ch
 
         if($check_fundation) {
-            if($fundation_id) {
+            if($fundation_id and $fundation_id != "NULL") {
```
Pourquoi ce `"NULL"` en tant que string ?

Pour être précis, PSR-2 recommande d'écrire `null` (en minuscules donc, par opposition au `NULL` MySQL). Dans tous les cas, cette ligne devrait vérifier l'identité comme ça : 
`if($fundation_id && $fundation_id !== NULL) {`

mattgu74 replied:
Je vais essayer de changer ça en modifiant aussi les "ALL" cf #95  Je ne sais pas si ça passera d'ici r8 mais au pire ça sera fait pour r9.

apuyou commented on class/ApplicationRight.class.php:
```diff
@@ -95,7 +103,11 @@ public static function getRights($fun_id) {
     public static function setRight($app_id, $service, $fun_id) {
         $db = Db_buckutt::getInstance();
-        $db->query("INSERT INTO tj_app_fun_afu (app_id, fun_id, afu_service) VALUES('%u', '%u', '%s');", Array($app_id, $fun_id, $service));
+        if($fun_id == "NULL") {
```
Même remarque, `if($fun_id == null)`

mattgu74 replied:
Même réponse cf #95

apuyou commented on class/ApplicationRight.class.php:
```diff
@@ -108,7 +120,11 @@ public static function setRight($app_id, $service, $fun_id) {
 	public static function removeRight($app_id, $service, $fun_id) {
         $db = Db_buckutt::getInstance();
-		$db->query("UPDATE tj_app_fun_afu SET afu_removed=NOW() WHERE fun_id='%u' AND app_id='%u' AND afu_service='%s';", Array($fun_id, $app_id, $service));
+        if($fun_id == "NULL") {
```
Idem.

mattgu74 replied:
#95

apuyou commented on class/UserRight.class.php:
```diff
@@ -96,15 +104,13 @@ public static function getRights($fun_id) {
-        $usr_id = $db->result($db->query("SELECT usr_id FROM ts_user_usr WHERE usr_nickname='%s';", Array($usr_login)),0);
-		if ($db->affectedRows() != 1) {
-			throw new Exception("Le login \"$usr_login\" n'a donné aucun resultat.");
-		}
```
Indentation loupée

mattgu74 replied:
En même temps tu commentes les lignes supprimés... C'était loupé avant. Maintenant les insertions au dessous en vert sont plutôt réussies, non ?

apuyou replied:
:bow:

mattgu74 committed:
le service dans les tables de droits peut être NULL

mattgu74 committed:
Réécriture de getServices dans ADMINRIGHT

mattgu74 committed:
Modification de getFundations dans ServiceBase (null ne peut etre une clef)

mattgu74 committed:
Correction de ApplicationRight pour gérer NULL comme étant le passe partout (au lieu de 'ALL' et 'NULL')

mattgu74 committed:
fix #95 'ALL' n'existe plus

mattgu74 committed:
Meilleurs gestion du blocage par les 'superadmins'

mattgu74 commented:
On peut savoir pourquoi cette PR n'est toujours pas merged ? J'aimerais bien bosser sur mes tickets pour R9, et ça implique de modifier les mêmes fichiers... Du coup pour éviter les merges foireux et les PR qui ressemble à rien parce que reprennent des commits déjà mergé ce serait bien de faire quelque chose ;)

mattgu74 commented:
PS: Travis est passé le problème est lié à une erreur de merde sur une des deux builds qui n'a rien a voir avec le code de payutc.

trecouvr commented:
Des petits unittests ça serait bien.

apuyou commented on class/ApplicationRight.class.php:
```diff
@@ -95,7 +103,31 @@ public static function getRights($fun_id) {
+            $query_end .= ", NULL";
+        }    
+
+        $db->query($query_start . $query_end . ");", $var);
```
Ce bloc de code est absolument pas clair. Quitte à avoir 4 cas, autant écrire les 4 requêtes différentes, au moins ce sera lisible.

mattgu74 replied:
Ou sinon on utilise un ORM ?

apuyou replied:
Je t'en prie.

apuyou commented on class/ApplicationRight.class.php:
```diff
@@ -108,7 +140,24 @@ public static function setRight($app_id, $service, $fun_id) {
+            $query .= "AND afu_service IS NULL ";
+        }  
+
+        $db->query($query, $var);
```
Même chose ici. De manière générale faut éviter de construire des requêtes comme ça, c'est impossible à maintenir.

mattgu74 replied:
Bah... On utilise un ORM ? Comment tu veux que je fasse, tu crois que ce serait mieux d'écrire 4 requêtes complètes, j'en suis pas convaincu...

apuyou commented on class/UserRight.class.php:
```diff
@@ -96,15 +104,33 @@ public static function getRights($fun_id) {
+            $query_end .= ", NULL";
+        }    
+
+        $db->query($query_start . $query_end . ");", $var);
```
Pareil

apuyou commented on class/UserRight.class.php:
```diff
@@ -117,7 +143,24 @@ public static function setRight($usr_login, $service, $fun_id) {
+            $query .= "AND ufu_service IS NULL ";
+        }  
+
+        $db->query($query, $var);
```
Encore :(

mattgu74 replied:
C'est bon c'est le dernier, y'en a que 4 ^^ (Insert et Update *2 (application et user))

trecouvr commented:
Sinon on peut se faire une classe construction de requette, ça prend un array pour le select, un array pour le where, un array pour le join, etc... ça colle le tout et ça renvoie le résultat. C'est tout con mais ça permet de rendre beaucoup de code plus clair.

mattgu74 commented:
Moé au final on se recode un orm (ptetre que orm n'est pas le bon nom, nous
se qu'on cherche c'est un constructeur de requete), parce qu'on a pas
réussi à en trouver un qui correspondait à nos besoins ?

mattgu74 committed:
Ajout d'une methode pour supprimer une image

mattgu74 committed:
ServiceBase: Correction d'une erreur dans getFundations + ajout d'une ...

méthode pour récupérer les images (Contrairement à la méthode dans poss2) elle respect
les ratio lors du redimenssionnement des images.

mattgu74 committed:
Ajout d'un Bom pour la gestion des Category

mattgu74 committed:
Ajout d'un Bom pour la gestion des Product

mattgu74 committed:
Ajout du service GESARTICLE

mattgu74 committed:
fix #26, changement de parent pour un article

mattgu74 committed:
fix #151, changement de prix pour un article

mattgu74 committed:
Merge branch 'master' into branche_2

trecouvr committed:
correct indentation

trecouvr committed:
Add some unittests

trecouvr committed:
Correct indentation

trecouvr committed:
Add some unittests for Product & Category

trecouvr commented:
J'ai corrigé l'indentation dans quelques fichiers et j'ai ajouté des quelques unittests de lecture seulement, c'est un bon début puisque le plus important pour moi c'est que ça merde pas sur la caisse, si ça merde à l'édition on pourra toujours modifier à la main dans la db.
C'est bon pour moi. Faudrait merge assez vite parce que c'est très gros là.

mattgu74 commented:
C'est bon pour moi aussi.
